File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/receivers/scrobble/PlaybackFileStoppedScrobblerRegistration.java
package com.lasthopesoftware.bluewater.client.playback.service.receivers.scrobble;


import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.client.connection.receivers.IConnectionDependentReceiverRegistration;
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.PlaylistEvents;

import java.util.Arrays;
import java.util.Collection;

public class PlaybackFileStoppedScrobblerRegistration implements IConnectionDependentReceiverRegistration {

	private static final Collection<IntentFilter> intents =
		Arrays.asList(
			new IntentFilter(PlaylistEvents.onPlaylistTrackComplete),
			new IntentFilter(PlaylistEvents.onPlaylistStop),
			new IntentFilter(PlaylistEvents.onPlaylistPause));

	@Override
	public BroadcastReceiver registerWithConnectionProvider(IConnectionProvider connectionProvider) {
		return new PlaybackFileStoppedScrobbleDroidProxy(ScrobbleIntentProvider.getInstance());
	}

	@Override
	public Collection<IntentFilter> forIntents() {
		return intents;
	}

	private static class PlaybackFileStoppedScrobbleDroidProxy extends BroadcastReceiver {

		private final ScrobbleIntentProvider scrobbleIntentProvider;

		public PlaybackFileStoppedScrobbleDroidProxy(ScrobbleIntentProvider scrobbleIntentProvider) {
			this.scrobbleIntentProvider = scrobbleIntentProvider;
		}

		@Override
		public void onReceive(Context context, Intent intent) {
			context.sendBroadcast(scrobbleIntentProvider.provideScrobbleIntent(false));
		}
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/receivers/scrobble/PlaybackFileStoppedScrobblerRegistration.kt
package com.lasthopesoftware.bluewater.client.playback.service.receivers.scrobble

import android.content.Context
import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider
import com.lasthopesoftware.bluewater.client.connection.receivers.RegisterReceiverForEvents
import com.lasthopesoftware.bluewater.client.playback.service.broadcasters.messages.PlaybackMessage
import com.lasthopesoftware.bluewater.shared.cls
import com.lasthopesoftware.bluewater.shared.messages.application.ApplicationMessage

class PlaybackFileStoppedScrobblerRegistration(private val context: Context) : RegisterReceiverForEvents {
	override fun registerWithConnectionProvider(connectionProvider: IConnectionProvider): (ApplicationMessage) -> Unit =
		PlaybackFileStoppedScrobbleDroidProxy(ScrobbleIntentProvider.getInstance())

	override fun forClasses(): Collection<Class<out ApplicationMessage>> = classes

	private inner class PlaybackFileStoppedScrobbleDroidProxy(private val scrobbleIntentProvider: ScrobbleIntentProvider) :
		(ApplicationMessage) -> Unit {

		override fun invoke(p1: ApplicationMessage) {
			context.sendBroadcast(scrobbleIntentProvider.provideScrobbleIntent(false))
		}
	}

	companion object {
		private val classes by lazy {
			setOf<Class<out ApplicationMessage>>(cls<PlaybackMessage.PlaybackStopped>(), cls<PlaybackMessage.TrackCompleted>())
		}
	}
}
